packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m

```diff
@@ -0,0 +1,188 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "GoogleMapHeatmapController.h"
+#import "JsonConversions.h"
+
+@implementation FLTGoogleMapHeatmapController {
+  GMUHeatmapTileLayer* _heatmapLayer;
+  GMSMapView* _mapView;
+}
+- (instancetype)initHeatmapWithHeatmapId:(NSString*)heatmapId mapView:(GMSMapView*)mapView {
+  self = [super init];
+  if (self) {
+    _heatmapLayer = [[GMUHeatmapTileLayer alloc] init];
+    _heatmapLayer.map = mapView;
+    _mapView = mapView;
+    _heatmapId = heatmapId;
+  }
+  return self;
+}
+
+- (void)removeHeatmap {
+  _heatmapLayer.map = nil;
+}
+
+#pragma mark - FLTGoogleMapHeatmapOptionsSink methods
+
+- (void)setVisible:(BOOL)visible {
+  _heatmapLayer.map = visible ? _mapView : nil;
+}
+- (void)setRadius:(NSUInteger)radius {
+  [_heatmapLayer setRadius:radius];
+  [_heatmapLayer clearTileCache];
+}
+- (void)setGradient:(GMUGradient*)gradient {
+  [_heatmapLayer setGradient:gradient];
+  [_heatmapLayer clearTileCache];
+}
+- (void)setPoints:(NSArray<GMUWeightedLatLng*>*)points {
+  [_heatmapLayer setWeightedData:points];
+  [_heatmapLayer clearTileCache];
+}
+- (void)setOpacity:(double)opacity {
+  _heatmapLayer.opacity = opacity;
+  [_heatmapLayer clearTileCache];
+}
+
+@end
+
+static int ToInt(NSNumber* data) { return [FLTGoogleMapJsonConversions toInt:data]; }
+
+static double ToDouble(NSNumber* data) { return [FLTGoogleMapJsonConversions toDouble:data]; }
+
+static BOOL ToBool(NSNumber* data) { return [FLTGoogleMapJsonConversions toBool:data]; }
+
+static NSArray<GMUWeightedLatLng*>* ToPoints(NSArray* data) {
+  NSMutableArray* points = [[NSMutableArray alloc] init];
+  for (unsigned i = 0; i < [data count]; i++) {
+    NSNumber* latitude = data[i][0][0];
+    NSNumber* longitude = data[i][0][1];
+    NSNumber* intensity = data[i][1];
```

Comment:
will any of these lines crash? For example, will `data[i].length < 2` and thus crashes `data[i][1]`

Reply:
That would most definitely crash, but that would also mean that the flutter part of this package is implemented improperly.
Should we throw an exception when the data is not correct, or how can this ideally be resolved?